environmental_data_workflows/post_processor.py

```diff
@@ -0,0 +1,634 @@
+        c) find and remove the synoptic spot readings (these are single readings per day - so we will
+                     identify all such single readings and remove them, if they are synoptic spot readings or not)
+        (Points (a) and (c) based on pers. comms. with Martyn Sunter, Met Office, July 2020.
+         Point (b) based on data exploration by authors.)
+    2)
+
+@author: mbessdl2
+"""
+
+import pandas as pd
+import numpy as np
+from datetime import datetime
+from abc import ABCMeta, abstractmethod
+import os
+try:
+    import metpy.calc as mpcalc
+    from metpy.units import units
+    import geopy.distance as distance
+    from sklearn.experimental import enable_iterative_imputer
+    from sklearn.impute import IterativeImputer
+    from sklearn.linear_model import BayesianRidge
+    from sklearn import preprocessing
+except ImportError as err:
+    print('Warning: Unable to load library: {}'.format(err))
+
+from environmental_data_workflows import EnvironmentWorkflow
+
+class PostProcessor(EnvironmentWorkflow):
+    __metaclass__ = ABCMeta
+
+    DEFAULT_IMPUTE_DATA = True
+    DEFAULT_PRINT_STATS = True
+    DEFAULT_FILE_IN = 'data_met/temp_rh_press_dewpoint_2016-2019.csv'
+    BASE_FILE_OUT = '{}/daily_mean_max_temp_RH_pres{}.csv'
+    DEFAULT_SKIP_INPUT_ROWS = 1
+
+
+    def __init__(self, out_dir=EnvironmentWorkflow.DEFAULT_OUT_DIR, verbose=EnvironmentWorkflow.DEFAULT_VERBOSE):
+        super(PostProcessor, self).__init__(out_dir, verbose)
+
+        self.impute_data = PostProcessor.DEFAULT_IMPUTE_DATA
+        self.print_stats = PostProcessor.DEFAULT_PRINT_STATS
+        self.file_in = PostProcessor.DEFAULT_FILE_IN
+        self._skip_input_rows = MetPostProcessor.DEFAULT_SKIP_INPUT_ROWS
+
+
+    #%% station geographic routines
+    def calc_station_distances(self, stations_in, stat_location):
+
+        station_distances = pd.DataFrame(index=stations_in.index)
+        station_distances['Distance'] = np.nan
+
+        for index, row in stations_in.iterrows():
+            new_location = (row['Latitude'],row['Longitude'])
+
+            station_distances.loc[index]['Distance'] = distance.distance(stat_location,new_location).km
+
+        return station_distances
+        return station_distances
+
+
+    # %% function for creating date objects
+    def parse_calcs_date(self, date_in):
+        return datetime.strptime(date_in, self._date_calcs_format)
+
+    def calc_nanmean(self, data_in):
+        return np.nanmean(data_in)
+
+
+class MetPostProcessor(PostProcessor):
+
+    DEFAULT_COLS_SPECIFIC_LIST = ['temperature', 'rel_hum', 'pressure', 'dewpoint']
+    DEFAULT_EXCLUDE_STATION_LIST = [117]
+    DEFAULT_MIN_TEMPERATURE = -20
+    DEFAULT_STATIONS_FILENAME = "../station_data/station_data_clean.csv"
+    DEFAULT_REFERENCE_NUMBER_STATIONS = 5
+    DEFAULT_REFERENCE_NUM_YEARS = 0.0625 #3.5
+    DEFAULT_MIN_YEARS = 0.04 #1
+
+    DEFAULT_DATE_CALCS_FORMAT = '%Y-%m-%d %H:%M:%S'
```

Comment:
Using this string in the help output causes a crash, as %Y (etc) are read as formatting strings. Can we strip the % out when we pass this to the help printout?